crates/mfem/src/lib.rs

```diff
@@ -183,54 +141,79 @@ pub enum RefAlgo {
     B = 1,
 }
 
-pub use mfem_sys::ffi::BasisType;
+pub use mfem_sys::BasisType;
 
 /////////////////////////////
 // FiniteElementCollection //
 /////////////////////////////
 
-pub trait FiniteElementCollection: AsBase<mfem_sys::ffi::FiniteElementCollection> {
+pub struct FiniteElementCollection {
+    inner: UniquePtr<mfem_sys::FiniteElementCollection>,
+}
+
+pub struct FiniteElementCollectionRef<'a> {
+    inner: &'a mfem_sys::FiniteElementCollection,
+}
+
+impl Deref for FiniteElementCollection {
+    type Target = FiniteElementCollectionRef<'a>;
+    fn deref(&self) -> &Self::Target {
+        todo!()
+    }
+}
+
+impl DerefMut for FiniteElementCollection {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        todo!()
+    }
+}
+
+impl<'a> FiniteElementCollectionRef<'a> {
     fn get_name(&self) -> String {
-        let ptr = self.as_base().Name();
+        let ptr = self.Name();
         assert!(!ptr.is_null());
         let name = unsafe { std::ffi::CStr::from_ptr(ptr) };
         name.to_owned().into_string().expect("Valid string")
     }
 }
```

Comment:
@Chris00 I need your input here! 😅 

I'm trying to work through this by hand so that I understand the gist of your technique before we descend into Macro Hell(TM).